### ISSUE TITLE
Can't get my head around the exact configuration requirements

### ISSUE THREAD

**Issue by wanno-drijfhout** [thomseddon/traefik-forward-auth#108]
Hi! I've been trying to get traefik-forward-auth working because I've been suffering from either redirect loops or 404s in my various bazooka-configuring attempts.

Today I got it to work, but I don't understand why and I think it's unnecessary convoluted. **My question to you:** would you help me understand/correct/improve my configuration?

I've taken inspiration from #79, #89, #81, #92 but I couldn't find the answer for my case there..

### Context

- Docker swarm (1 Linux node)
- Only using OIDC-provider: Keycloak
- Traefik 2.2
  - Default host rule: 
    ```
    - --providers.docker.defaultRule=Host(`{{ normalize (index .Labels "com.docker.stack.namespace") }}.$MYDOMAIN`)
    ```
  - Redirect HTTP to HTTPS; certificates managed by letsencrypt

### The problem

I finally got it to work with this little piece of configuration. By the way, I also have to respecify the middleware for all my Docker stack files.

```yml
    hostname: forward-auth
    ports:
      - 4181:4181
# ...
        - "traefik.http.middlewares.forward-auth-verify.forwardauth.address=http://forward-auth:4181"
```

This means I'm circumventing Traefik for the forward-authentication. I don't understand why I'd have to. Ideally, I'd write this:
```yml
#    hostname: <unspecified, so random>
#    ports:
#      - 4181:4181 <unspecified>
# ...
        - "traefik.http.middlewares.forward-auth-verify.forwardauth.address=https://account.$MYDOMAIN/login"  # so, at httpS, reverse-proxied via traefik
```

### Account stack

This stack includes Keycloak and the traefik-forward-auth services:

```yml
version: '3.6'

services:
  # The forward-auth is the "auth host" and must be configured as "client" in keycloak.
  # "when a user returns from authentication with a 3rd party provider they will always be forwarded to this [auth] host."
  forward-auth:
    image: thomseddon/traefik-forward-auth:2.1
    command:
      - --log-level=debug
      # We include a separate config file for secret properties:
      #   - --providers.oidc.client-id="..."
      #   - --providers.oidc.client-secret="..."
      #   - --secret="..."
      - --config=/run/secrets/FORWARD_AUTH_SECRETS
      - --default-provider=oidc
      # Check https://account.$MYDOMAIN/auth/realms/master/.well-known/openid-configuration
      # The issuer must refer to the keycloak below
      - --providers.oidc.issuer-url=https://account.$MYDOMAIN/auth/realms/master
      # The auth host must refer to this forward-auth
      - --auth-host=account.$MYDOMAIN
      - --cookie-domain=$MYDOMAIN
      #- --domain=$MYDOMAIN
      # After authentication, we're routed back to this auth host
      - --url-path=/_oauth
      # To disable HTTPS (for debugging purposes), you may consider: (do not use in production!)
      #- --insecure-cookie
    secrets:
      - FORWARD_AUTH_SECRETS
    depends_on:
      - keycloak
    networks:
      - proxy
    hostname: forward-auth
    ports:
      - 4181:4181
    deploy:
      replicas: 1
      labels:
        - "traefik.http.services.forward-auth.loadbalancer.server.port=4181"
        # Forward auth goes over internal network instead of regular 'proxy' network
        - "traefik.http.middlewares.forward-auth-verify.forwardauth.address=http://forward-auth:4181"
        - "traefik.http.middlewares.forward-auth-verify.forwardauth.trustForwardHeader=true"
        - "traefik.http.middlewares.forward-auth-verify.forwardauth.authResponseHeaders=X-Forwarded-User"
        #- "traefik.http.middlewares.redirect-login.redirectregex.regex=^https?://account.$MYDOMAIN/login"
        #- "traefik.http.middlewares.redirect-login.redirectregex.replacement=http://forward-auth:4181"
        - "traefik.http.middlewares.redirect-logout.redirectregex.regex=^https?://account.$MYDOMAIN/logout"
        - "traefik.http.middlewares.redirect-logout.redirectregex.replacement=https://account.$MYDOMAIN/_oauth/logout"
        - "traefik.http.routers.forward-auth.middlewares=forward-auth-verify,redirect-logout"
        - "traefik.http.routers.forward-auth.tls.certresolver=letsencrypt"
        - "traefik.http.routers.forward-auth.entrypoints=http,https"
        - "traefik.http.routers.forward-auth.rule=Host(`account.$MYDOMAIN`) &&( PathPrefix(`/login`) || PathPrefix(`/logout`) || PathPrefix(`/_oauth`) )"
  
  # keycloak is the "token issuer/provider" and account manager
  keycloak:
    image: quay.io/keycloak/keycloak:latest
    environment:
      KEYCLOAK_USER_FILE: /run/secrets/KEYCLOAK_USER
      KEYCLOAK_PASSWORD_FILE: /run/secrets/KEYCLOAK_PASSWORD
      #KEYCLOAK_FRONTEND_URL: http://account.$MYDOMAIN/auth/ ???
      # KEYCLOAK_LOGLEVEL: INFO (ALL, DEBUG, ERROR, FATAL, INFO, OFF, TRACE or WARN)
      # KEYCLOAK_WELCOME_THEME
      # KEYCLOAK_DEFAULT_THEME

      DB_VENDOR: POSTGRES
      # DB_ADDR should be the name of the Postgres service define further down the file
      DB_ADDR: keycloak_db
      DB_DATABASE: keycloak
      DB_SCHEMA: public
      DB_USER_FILE: /run/secrets/KEYCLOAK_DB_USER
      DB_PASSWORD_FILE: /run/secrets/KEYCLOAK_DB_PASSWORD
      
      # Uncomment the line below if you want to specify JDBC parameters. The parameter below is just an example, and it shouldn't be used in production without knowledge. It is highly recommended that you read the PostgreSQL JDBC driver documentation in order to use it.
      #JDBC_PARAMS: "ssl=true"

      # This is required to run keycloak behind traefik
      PROXY_ADDRESS_FORWARDING: "true"
    #ports:
    #  - 8080:8080
    networks:
     - default
     - proxy
    depends_on:
     - keycloak_db
    secrets:
     - KEYCLOAK_USER
     - KEYCLOAK_PASSWORD
     - KEYCLOAK_DB_USER
     - KEYCLOAK_DB_PASSWORD
    deploy:
      replicas: 1
      labels:
        - "traefik.http.services.keycloak-app.loadbalancer.server.port=8080"
        - "traefik.http.middlewares.redirect-accountconsole.redirectregex.regex=^https?://account.$MYDOMAIN/me/?"
        - "traefik.http.middlewares.redirect-accountconsole.redirectregex.replacement=https://account.$MYDOMAIN/auth/realms/master/account/"
        - "traefik.http.routers.keycloak-app.middlewares=redirect-accountconsole"
        - "traefik.http.routers.keycloak-app.tls.certresolver=letsencrypt"
        - "traefik.http.routers.keycloak-app.entrypoints=http,https"
        - "traefik.http.routers.keycloak-app.rule=Host(`account.$MYDOMAIN`)"

  keycloak_db:
    image: postgres
    networks:
     - default
    #ports:
    # - "5432:5432"
    environment:
      POSTGRES_DB: keycloak
      POSTGRES_USER_FILE: /run/secrets/KEYCLOAK_DB_USER
      POSTGRES_PASSWORD_FILE: /run/secrets/KEYCLOAK_DB_PASSWORD
    volumes:
     - keycloak_db:/var/lib/postgresql/data
    secrets:
     - KEYCLOAK_DB_USER
     - KEYCLOAK_DB_PASSWORD

volumes:
  keycloak_db:

secrets:
  FORWARD_AUTH_SECRETS:
    external: true
  KEYCLOAK_USER:
    external: true
  KEYCLOAK_PASSWORD:
    external: true
  KEYCLOAK_DB_USER:
    external: true
  KEYCLOAK_DB_PASSWORD:
    external: true


networks:
  default:
  proxy:
    external: true
```

### Whoami stack (requires authentication; test)

```yml
version: '3.6'

services:
  app:
    image: containous/whoami
    networks:
     - proxy
    deploy:
      mode: global
      labels:
        - "traefik.http.services.whoami-app.loadbalancer.server.port=80"
        - "traefik.http.middlewares.forward-auth.forwardauth.address=http://forward-auth:4181"
        - "traefik.http.middlewares.forward-auth.forwardauth.trustForwardHeader=true"
        - "traefik.http.middlewares.forward-auth.forwardauth.authResponseHeaders=X-Forwarded-User"
        - "traefik.http.routers.whoami-app.middlewares=forward-auth"
        - "traefik.http.routers.whoami-app.tls.certresolver=letsencrypt"
        - "traefik.http.routers.whoami-app.entrypoints=https"

networks:
  proxy:
    name: proxy
    external: true
```

**Comment by SuperSandro2000** [comment]
If you take a look at the graphic at the top of https://docs.traefik.io/middlewares/forwardauth/ you will see that traefik does an internal connection to traefik-forward-auth.

This means if you send it to https://auth.domain.com it will open a connection to the external service which means you end up in an authentication loop.

**Comment by DirtyCajunRice** [comment]
So to hop in on this for the internal vs external... You cannot host the forward auth somewhere else and reach it externally? Why not? Does this mean you need to host a traefik-forward-auth in every environment and keep the secret the same?

**Comment by wanno-drijfhout** [comment]
> If you take a look at the graphic at the top of https://docs.traefik.io/middlewares/forwardauth/ you will see that traefik does an internal connection to traefik-forward-auth.

I fear I do not completely understand. How does that picture imply that traefik does an _internal connection_ to fraefik-forward-auth and what does that mean?

> This means if you send it to https://auth.domain.com it will open a connection to the external service which means you end up in an authentication loop.

But if traefik-forward-auth is running on auth.domain.com, why can't it handle the request, then? Is there a difference in HTTP requests as traefik-forward-auth would receive them:
- from the "AuthForward" middleware directly to the fraefik-forward-auth via Docker-exposed port 4181, or;
- from the "AuthForward" middleware by just forwarding to "auth.domain.com" via Traefik "Routers", or;
- from handling the "Service" endpoint (loadbalanced on port 4181).

I do see the potential for recursive loops in the second option, as you say, but could only be the case if:
- I define "AuthForward" middleware on the traefik-forward-auth service, and;
- traefik-forward-auth doesn't detect that it's being asked to forward authentication to itself.

Question: Do I really need to define "AuthForward" middleware on the traefik-forward-auth service (like I've done in my configuration above)?

**Comment by thomseddon** [comment]
Hi, yes of course we'd be happy to help!

As @SuperSandro2000 says - the architecture of forward authentication in traefik is that for every request that is received, it will forward a copy to your desired destination - in this case traefik-forward-auth .

If traefik-forward-auth finds a valid session in the request, it will return a 200 OK to traefik. When traefik receives this it will proceed with proxying the request to your desired back end as usual.

If traefik-forward-auth does not find a valid session, it will return a non 200 response (either a redirect to your auth provider, or an error if something went wrong). When traefik receives this it will not proceed with proxying the request to your desired back end, but will instead pass this response directly to the user.

So traefik-forward-auth does not proxy the request, is just there to let traefik know if the request should or shouldn't be allowed. And if it isn't allowed, what to do.
When you had `https://account.$MYDOMAIN/login"` as the forward authentication location, this would actually be forwarded to traefik again, which would in turn create another forward authentication lookup and just keep looping.

So, in summary - the config you have ended up with seems correct for traefik v2. I am currently adding lots more examples for traefik v2 for both swarm and k8s to help simplify things.

**Comment by wanno-drijfhout** [comment]
Thank you so much!

> So traefik-forward-auth does not proxy the request, is just there to let traefik know if the request should or shouldn't be allowed. And if it isn't allowed, what to do.
> When you had `https://account.$MYDOMAIN/login"` as the forward authentication location, this would actually be forwarded to traefik again, which would in turn create another forward authentication lookup and just keep looping.

Would it be possible for traefik-forward-auth to always allow requests to itself, then? This could break the infinite loop, if I understand correctly

**Comment by thomseddon** [comment]
So your current configuration, where the forward authentication requests are sent directly from traefik to traefik-forward-auth on port 4181, is the correct and intended design and you should not see an infinite loop.

I apologise  that there aren't any kubernetes examples in the repo - I have added 3 in my local version and I'm just building out the docs so it's as clear as possible.

I'm going to close this one for now, but please feel free to add any more questions :)
